dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/trackedentity/job/TrackerTrigramIndexingJobTest.java

```diff
@@ -0,0 +1,140 @@
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.trackedentity.job;
+
+import static org.mockito.Mockito.*;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.hisp.dhis.scheduling.JobConfiguration;
+import org.hisp.dhis.scheduling.NoopJobProgress;
+import org.hisp.dhis.scheduling.parameters.TrackerTrigramIndexJobParameters;
+import org.hisp.dhis.trackedentity.TrackedEntityAttribute;
+import org.hisp.dhis.trackedentity.TrackedEntityAttributeService;
+import org.hisp.dhis.trackedentityattributevalue.TrackerTrigramIndexingJob;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * Tests the {@link TrackerTrigramIndexingJob} using mocks to only test the
+ * logic of the job.
+ *
+ * @author Ameen
+ */
+public class TrackerTrigramIndexingJobTest
+{
+
+    private final TrackedEntityAttributeService trackedEntityAttributeService = mock(
+        TrackedEntityAttributeService.class );
+
+    private final TrackerTrigramIndexingJob job = new TrackerTrigramIndexingJob( trackedEntityAttributeService );
+
+    @Before
+    public void setUp()
+    {
+        // mock normal run conditions
+        when( trackedEntityAttributeService.getAllTrigramIndexableTrackedEntityAttributes() ).thenReturn(
+            Collections.emptySet() );
+        when( trackedEntityAttributeService.runAnalyze() ).thenReturn( true );
+        when( trackedEntityAttributeService.runVacuum() ).thenReturn( true );
+    }
+
+    @Test
+    public void testRunJobWithoutAnyAttributesInJobParameters()
+    {
+        JobConfiguration jobConfiguration = new JobConfiguration();
+        TrackerTrigramIndexJobParameters jp = new TrackerTrigramIndexJobParameters();
+        jp.setRunAnalyze( true );
+        jp.setRunVacuum( true );
+        jobConfiguration.setJobParameters( jp );
+
+        job.execute( jobConfiguration, NoopJobProgress.INSTANCE );
+
+        verify( trackedEntityAttributeService, atLeastOnce() ).runAnalyze();
```

Comment:
why is this one checking atLeastOnce and the other tests atMostOnce?

Reply:
I will change everything to make it more consistent. Since I know the exact numbers, I think I can avoid using both atLeast and atMost and test for the exact numbers instead. Thanks!